app/code/Magento/InventoryCatalog/Plugin/CatalogInventory/Model/ResourceModel/UpdateSourceItemAtLegacyCatalogInventoryQtyCounter.php

```diff
@@ -0,0 +1,147 @@
+<?php
+/**
+ * Copyright © Magento, Inc. All rights reserved.
+ * See COPYING.txt for license details.
+ */
+declare(strict_types=1);
+
+namespace Magento\InventoryCatalog\Plugin\CatalogInventory\Model\ResourceModel;
+
+use Magento\Catalog\Api\ProductRepositoryInterface;
+use Magento\Framework\App\ResourceConnection;
+use Magento\InventoryApi\Api\Data\SourceItemInterface;
+use Magento\InventoryApi\Api\SourceItemsSaveInterface;
+use Magento\InventoryApi\Api\SourceItemRepositoryInterface;
+use Magento\Framework\Api\SearchCriteriaBuilder;
+use Magento\InventoryCatalog\Api\DefaultSourceProviderInterface;
+
+/**
+ * Class provides around Plugin on Magento\CatalogInventory\Model\ResourceModel::correctItemsQty
+ * to update data in Inventory source item
+ */
+class UpdateSourceItemAtLegacyCatalogInventoryQtyCounter
+{
+    /**
+     * @var ProductRepositoryInterface
+     */
+    private $productRepository;
+
+    /**
+     * @var SourceItemRepositoryInterface
+     */
+    private $sourceItemRepository;
+
+    /**
+     * @var SourceItemsSaveInterface
+     */
+    private $sourceItemsSave;
+
+    /**
+     * @var DefaultSourceProviderInterface
+     */
+    private $defaultSourceProvider;
+
+    /**
+     * @var SearchCriteriaBuilder
+     */
+    private $searchCriteriaBuilder;
+
+    /**
+     * @var ResourceConnection
+     */
+    private $resourceConnection;
+
+    /**
+     * @param ProductRepositoryInterface $productRepository
+     * @param SourceItemRepositoryInterface $sourceItemRepository
+     * @param SourceItemsSaveInterface $sourceItemsSave
+     * @param DefaultSourceProviderInterface $defaultSourceProvider
+     * @param SearchCriteriaBuilder $searchCriteriaBuilder
+     * @param ResourceConnection $resourceConnection
+     */
+    public function __construct(
+        ProductRepositoryInterface $productRepository,
+        SourceItemRepositoryInterface $sourceItemRepository,
+        SourceItemsSaveInterface $sourceItemsSave,
+        DefaultSourceProviderInterface $defaultSourceProvider,
+        SearchCriteriaBuilder $searchCriteriaBuilder,
+        ResourceConnection $resourceConnection
+    ) {
+        $this->productRepository = $productRepository;
+        $this->sourceItemRepository = $sourceItemRepository;
+        $this->sourceItemsSave = $sourceItemsSave;
+        $this->defaultSourceProvider = $defaultSourceProvider;
+        $this->searchCriteriaBuilder = $searchCriteriaBuilder;
+        $this->resourceConnection = $resourceConnection;
+    }
+
+    /**
+     * @param $subject
+     * @param callable $proceed
+     * @param int[] $items
+     * @param int $websiteId
+     * @param string $operator +/-
+     *
+     * @return void
+     * @throws \Exception
+     * @SuppressWarnings(PHPMD.UnusedFormalParameter)
+     */
+    public function aroundCorrectItemsQty($subject, callable $proceed, array $items, $websiteId, $operator)
+    {
+        $proceed($items, $websiteId, $operator);
```

Comment:
we need to wrap with transaction initial method as well 
so, 
```
   $connection = $this->resourceConnection->getConnection();
 +        $connection->beginTransaction();
```
should be before $proceed